### PR TITLE
BENCH: cKDTree benchmarks added: balanced/unbalanced tree (related to PR-11653)

### PR DESCRIPTION
**Pull request by scidam**
@rgommers, @sturlamolden 

Hello, 
I've prepared benchmarks for building and querying balanced/unbalanced kd-trees.
Also, I performed comparison with old (partial quick sort)  algorithm on presorted data (results are below).  The `nth_element`-based version works faster.

#### Reference issue
Related issue #11595
Previous PR: #11653

#### Additional information
These are some results I've got:
```
Building trees: New partial_sort.h
              --                                           balanced / order                        
              ------------------- -----------------------------------------------------------------
                   (m, n, r)       True / random   True / sorted   False / random   False / sorted 
              =================== =============== =============== ================ ================
                (3, 10000, 1000)     5.45±0.1ms     3.11±0.03ms     3.42±0.05ms      2.20±0.02ms   
                (8, 10000, 1000)     6.47±0.5ms      3.94±0.2ms     4.24±0.07ms      3.02±0.03ms   
               (16, 10000, 1000)     9.44±0.3ms     5.63±0.05ms      6.98±0.2ms      4.53±0.07ms   
              =================== =============== =============== ================ ================


Old partial_sort.h
                                                                balanced / order
              ------------------- -----------------------------------------------------------------
                   (m, n, r)       True / random   True / sorted   False / random   False / sorted 
              =================== =============== =============== ================ ================
                (3, 10000, 1000)     5.82±0.1ms       340±9ms       3.53±0.06ms      2.28±0.02ms   
                (8, 10000, 1000)     6.60±0.4ms       365±20ms      4.32±0.06ms      3.09±0.05ms   
               (16, 10000, 1000)      11.3±2ms        541±10ms      7.05±0.07ms      4.72±0.09ms   
              =================== =============== =============== ================ ================
```

### CODE REVIEW

File: benchmarks/benchmarks/spatial.py [hunk 1]
```diff
@@ -54,6 +54,61 @@ def time_build(self, mnr, cls_name):
...
+        super().setup(*args, None)
+
+    def time_build(self, mnr, balanced, order):
+        cKDTree(self.data.get(order), balanced_tree=balanced)
+
+
+class QueryUnbalanced(PresortedDataSetup):
+
+    def setup(self, *args):
+        super().setup(*args, None)
```
Comment (rgommers) [inline benchmarks/benchmarks/spatial.py]: Something is going wrong here:
```
· Discovering benchmarks
· Running 4 total benchmarks (1 commits * 1 environments * 4 benchmarks)
[  0.00%] ·· Benchmarking existing-py_home_rgommers_anaconda3_bin_python
[ 12.50%] ··· Running (spatial.BuildUnbalanced.time_build--)....
[ 62.50%] ··· spatial.BuildUnbalanced.time_build                                      ok
[ 62.50%] ··· =================== ========== ============= =============
              --                                        order           
              ------------------------------ ---------------------------
                   (m, n, r)       balanced      random        sorted   
              =================== ========== ============= =============
                (3, 10000, 1000)     True     2.28±0.03ms   1.29±0.01ms 
                (3, 10000, 1000)    False     1.31±0.02ms     874±10μs  
                (8, 10000, 1000)     True     2.79±0.04ms   1.75±0.02ms 
                (8, 10000, 1000)    False     1.77±0.05ms   1.27±0.03ms 
               (16, 10000, 1000)     True     3.34±0.07ms   2.23±0.01ms 
               (16, 10000, 1000)    False     2.22±0.01ms   1.79±0.02ms 
              =================== ========== ============= =============

[ 75.00%] ··· spatial.QueryUnbalanced.time_query                                  failed
[ 75.00%] ··· =================== ========== ============== ==============
              --                                     order / radius       
              ------------------------------ -----------------------------
                   (m, n, r)       balanced   random / 0.5   sorted / 0.5 
              =================== ========== ============== ==============
                (3, 10000, 1000)     True        failed         failed    
                (3, 10000, 1000)    False        failed         failed    
                (8, 10000, 1000)     True        failed         failed    
                (8, 10000, 1000)    False        failed         failed    
               (16, 10000, 1000)     True        failed         failed    
               (16, 10000, 1000)    False        failed         failed    
              =================== ========== ============== ==============

[ 75.00%] ···· For parameters: (3, 10000, 1000), True, 'random', 0.5
               Traceback (most recent call last):
                 File "/home/rgommers/anaconda3/lib/python3.7/site-packages/asv/benchmark.py", line 1184, in main_run_server
                   main_run(run_args)
                 File "/home/rgommers/anaconda3/lib/python3.7/site-packages/asv/benchmark.py", line 1052, in main_run
                   skip = benchmark.do_setup()
                 File "/home/rgommers/anaconda3/lib/python3.7/site-packages/asv/benchmark.py", line 583, in do_setup
                   result = Benchmark.do_setup(self)
                 File "/home/rgommers/anaconda3/lib/python3.7/site-packages/asv/benchmark.py", line 515, in do_setup
                   setup(*self._current_params)
                 File "/home/rgommers/code/scipy/benchmarks/benchmarks/spatial.py", line 100, in setup
                   super().setup(*args, None)
               TypeError: setup() takes 5 positional arguments but 6 were given
...               
```

This setup looks the same as the one for `BuildUnbalanced`, but `params`, `param_names` treatment is different.

**Comment by rgommers** [comment]
Thanks @scidam! This looks pretty good to me. If you could fix the failure, then I think it's about ready to merge.

**Comment by scidam** [comment]
@rgommers, thats was my fault; I fixed the issue and rebased the branch to the actual state (now travis-ci should be passing).

**Comment by rgommers** [comment]
LGTM now, merged. Thanks @scidam!

**Pull request merged**